tests/late_joiners.py

```diff
@@ -0,0 +1,216 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache 2.0 License.
+import os
+import getpass
+import time
+import logging
+import multiprocessing
+import shutil
+from random import seed
+import infra.ccf
+import infra.proc
+import infra.notification
+import infra.net
+import e2e_args
+from threading import Timer
+import random
+import contextlib
+
+from loguru import logger as LOG
+
+# 256 is the number of most recent messages that PBFT keeps in memory before needing to replay the ledger
+TOTAL_REQUESTS = 56
+
+
+def timeout(node, suspend, election_timeout):
+    if suspend:
+        # We want to suspend the nodes' process so we need to initiate a new timer to wake it up eventually
+        node.suspend()
+        next_timeout = random.uniform(2 * election_timeout, 3 * election_timeout)
+        LOG.info(f"New timer set for node {node.node_id} is {next_timeout} seconds")
+        t = Timer(next_timeout, timeout, args=[node, False, 0])
+        t.start()
+    else:
+        node.resume()
+
+
+def run(args):
+    hosts = ["localhost", "localhost", "localhost"]
+
+    with infra.ccf.network(
+        hosts, args.build_dir, args.debug_nodes, args.perf_nodes, pdb=args.pdb
+    ) as network:
+        network.start_and_join(args)
+        first_node, backups = network.find_nodes()
+
+        term_info = {}
+        long_msg = "X" * (2 ** 14)
+
+        nodes_to_kill = []
+        for b in backups:
+            nodes_to_kill.append(b)
```

Comment:
`nodes_to_kill = backups.copy()`

Reply:
Actually I realised we need 4 nodes to guarantee progress so I should only kill one of the backups and keep the other